src/spatial/hop_rates.jl

```diff
@@ -0,0 +1,174 @@
+"""
+A file with structs and functions for sampling hops and updating hopping rates
+"""
+
+### spatial hop rates ###
+abstract type AbstractHopRates end
+"""
+    update_hop_rates!(hop_rates::AbstractHopRates, species::AbstractArray, u, site, spatial_system)
+
+update rates of all specs in species at site
+"""
+function update_hop_rates!(hop_rates::AbstractHopRates, species::AbstractArray, u, site, spatial_system)
+    @inbounds for spec in species
+        update_hop_rate!(hop_rates, spec, u, site, spatial_system)
+    end
+end
+
+"""
+    total_site_hop_rate(hop_rates::AbstractHopRates, site)
+    
+return total hopping rate out of site
+"""
+total_site_hop_rate(hop_rates::AbstractHopRates, site) = @inbounds hop_rates.sum_rates[site]
+
+############## hopping rates of form L_{s,i} ################
+
+HopRates(hopping_constants::Matrix{F}) where F <: Number = HopRatesUnifNbr(hopping_constants)
+HopRates(hopping_constants::AbstractArray) where F <: Number = HopRatesGeneral(hopping_constants)
+
+struct HopRatesUnifNbr{F} <: AbstractHopRates
+    "hopping_constants[i,j] is the hop constant of species i at site j"
+    hopping_constants::Matrix{F} 
+
+    "rates[i,j] is total hopping rate of species i at site j"
+    rates::Matrix{F}
+
+    "sum_rates[j] is the sum of hopping rates at site j"
+    sum_rates::Vector{F}
+end
+
+function Base.show(io::IO, hop_rates::HopRatesUnifNbr)
```

Comment:
This is the wrong show dispatch. This probably should be for the text mimetype. 2-argument should is supposed to be round-trippable.

Reply:
Thanks. Fixed